### ISSUE TITLE
Metal 2.0 base ProgramSpec cache-hit path silently mis-binds aliased input tensors

### ISSUE THREAD

**Issue by gajanan-choudhary** [tenstorrent/tt-metal#57248]
## Summary

The Metal 2.0 **base `ProgramSpecFactoryConcept`** cache-hit path silently mis-binds **aliased input tensors**: if a program-cache entry is primed by a call that passes the same tensor handle in two argument slots, a later call of the same spec with two *distinct* tensors reuses the entry and reads the first tensor twice — wrong results, no error. The `CustomProgramSpecFactoryConcept` path is not affected. Every base-concept multi-input ported op is structurally exposed.

Found while porting `ttnn.concat` (#54651); concat has an op-local workaround, but the root cause is in the shared spec adapter and should be fixed there.

## Symptom (reproduced on Wormhole for concat)

```python
out1 = ttnn.concat([x, x], dim=0)   # primes the cache entry
out2 = ttnn.concat([a, b], dim=0)   # same spec -> cache hit -> returns concat(a, a), not concat(a, b)
```

`concat([x, y, x])` then `concat([a, b, c])` returns `[a, b, a]` — only the aliased slot goes stale. The reverse order (distinct first) is correct, so whether a process is affected depends on call order. Concat is a likely trigger: `repeat_interleave` passes one handle many times; diffusion guidance does `concat([latents, latents])`.

## Root cause

`SpecBindingSupport::resolve_bindings` ([`ttnn/api/ttnn/mesh_device_operation_adapter.hpp:892`](https://github.com/tenstorrent/tt-metal/blob/main/ttnn/api/ttnn/mesh_device_operation_adapter.hpp)) maps each `TensorArgument` to a position by `std::find_if` on the `MeshTensor` address and takes the **first** match. When the same handle appears twice, both parameters resolve to index 0, and that index table is frozen into the cache entry. On a cache hit, `refresh_tensor_args` (`:826`) replays the frozen table via `UpdateTensorArgs`, so a later distinct call rebinds `input_1` from index 0 as well. Nothing catches it: the only hit-time check compares tensor specs (which match), and it is off by default.

**Why `CustomProgramSpecFactoryConcept` is immune:** its hit path (`:1040`) re-runs the factory's `override_runtime_arguments` against the *current* tensors each dispatch and applies fresh `ProgramRunArgs` via `UpdateProgramRunArgs` — it never uses the frozen table, so aliasing is re-derived correctly every time.

**Legacy parity:** the legacy `ProgramDescriptor` patcher already guards this — [`tt_metal/impl/program/program_descriptor_patching.cpp:53-103`](https://github.com/tenstorrent/tt-metal/blob/main/tt_metal/impl/program/program_descriptor_patching.cpp) bails to a full rebuild when a buffer repeats within the inputs (the `matmul(X, X)` comment). The base spec path dropped that guarantee.

## Scope — exposed merged ops

Any base-concept (`create_program_artifacts`, no `override_runtime_arguments`) op that binds ≥2 input tensors that can be the same handle. From a scan of ported factories, this includes at least: `binary_ng`, `bcast`, `repeat_and_interleave_eltwise_mul`, `gelu_backward`, `scatter`, `indexed_fill`, `embedding`, `layernorm`/`layernorm_distributed`, `batch_norm`, `softmax`, `rotary_embedding_llama`, `sdpa`/`sdpa_decode`/`joint_sdpa`, `moreh_matmul`, `moreh_dot`(`_backward`), and several `moreh_*_backward`. (Structural exposure via the shared path; only concat has been reproduced on device.)

## Proposed fixes

**Option A — fold an input-aliasing signature into the program hash, framework-wide.** Aliased and distinct patterns never share an entry, so no bail-out/rebuild is needed. This generalizes concat's op-local workaround to one place. *Lower risk, smaller diff* — the main question is finding the single chokepoint where the base spec path computes the program hash so default-hash ops are covered too. Cost is a behavior change (aliased calls get extra cache entries) plus a cross-op regression test.

**Option B — restore the legacy bail-out in the spec adapter.** Detect duplicate indices in `resolve_bindings`, mark the entry, and on a cache hit take a *rebuild* branch (re-run `create_program_artifacts` → `MakeMeshWorkloadFromSpecs` → `SetProgramRunArgs`) instead of `UpdateTensorArgs`. *Higher effort* — the base spec hit-path has **no rebuild capability today** ("no Program rebuild", `:826` design comment), so that must be added; it also touches the correctness-critical hit path shared by every spec-ported op and should be gated by a parity check (the descriptor path uses `TT_DESCRIPTOR_PATCHING_PARITY_CHECK`; the spec path has no equivalent yet).

Recommendation: **Option A** as the lower-risk fix that also retroactively protects the merged ops above; Option B if the team prefers mirroring the legacy contract exactly. Metalium's call — this is their code.

## Current mitigation

`ttnn.concat` carries an op-local fix on `akertesz/concat-port-test` (commit `251ff951218`): its `compute_program_hash` folds an aliasing signature so aliased/distinct patterns get separate entries. Regression test: `tests/ttnn/unit_tests/operations/data_movement/test_concat_cache_aliasing.py`. Once the framework fix lands, that op-local workaround should be reverted.

## Acceptance criteria

- A base-concept multi-input spec-ported op, primed with an aliased call, returns correct data on a later distinct-input call of the same spec (and vice-versa).
- A regression test covering the alias-then-distinct and distinct-then-alias orderings for a base-concept multi-input op.
- No behavior change for `CustomProgramSpec` ops (already correct).

Related: #54651 (concat port where this surfaced), #55370 (concat Metal 2.0 sub-issue).
